components/org.wso2.identity.webhook.wso2.event.handler/src/main/java/org/wso2/identity/webhook/wso2/event/handler/api/builder/WSO2RegistrationEventPayloadBuilder.java

```diff
@@ -144,4 +166,76 @@ private List<UserClaim> filterUserClaimsForUserAdd(Map<String, String> userClaim
+        String tenantDomain = String.valueOf(properties.get(IdentityEventConstants.EventProperty.TENANT_DOMAIN));
+
+        String userStoreDomainName = resolveUserStoreDomain(properties, tenantDomain);
+        UserStore userStore = null;
+
+        if (StringUtils.isNotBlank(userStoreDomainName)) {
+            userStore = new UserStore(userStoreDomainName);
+        }
+
+        User newUser = new User();
+        enrichUser(properties, newUser);
+
+        Organization organization = new Organization(tenantId, tenantDomain);
+        Flow flow = IdentityContext.getThreadLocalIdentityContext().getFlow();
+        String initiatorType = null;
+        String action = null;
+        if (flow != null) {
+            initiatorType = flow.getInitiatingPersona().name();
+            action = flow.getName().name();
+        }
+
+        String errorCode = String.valueOf(properties.get(IdentityEventConstants.EventProperty.ERROR_CODE));
+        String errorMessage = String.valueOf(properties.get(IdentityEventConstants.EventProperty.ERROR_MESSAGE));
+
+        Context context = null;
+
+        if (properties.get(IdentityEventConstants.EventProperty.STEP_ID) != null) {
+
+            String idpName = String.valueOf(properties.get(IdentityEventConstants.EventProperty.IDP));
+            String currentAuthenticator =
+                    String.valueOf(properties.get(IdentityEventConstants.EventProperty.CURRENT_AUTHENTICATOR));
+            int stepId = Integer.parseInt((String) properties.get(IdentityEventConstants.EventProperty.STEP_ID));
+
+            Step failedStep = new Step(stepId, idpName, currentAuthenticator);
+            context = new Context(failedStep);
+
+        }
+
+        Reason reason = new Reason(errorCode, errorMessage, context);
+
+        return new WSO2RegistrationFailureEventPayload.Builder()
+                .initiatorType(initiatorType)
+                .action(action)
+                .user(newUser)
+                .tenant(organization)
+                .userStore(userStore)
+                .reason(reason)
+                .build();
+    }
+
+    private static String resolveUserStoreDomain(Map<String, Object> properties, String tenantDomain) {
+
+        String userStoreDomainName = null;
+        if (properties.get(IdentityEventConstants.EventProperty.USER_STORE_DOMAIN) != null) {
+            userStoreDomainName =
+                    String.valueOf(properties.get(IdentityEventConstants.EventProperty.USER_STORE_DOMAIN));
+        } else {
+            RealmConfiguration realmConfiguration =
+                    WSO2PayloadUtils.getRealmConfigurationByTenantDomain(tenantDomain);
+            if (realmConfiguration != null) {
+                userStoreDomainName =
+                        realmConfiguration.getUserStoreProperty(UserCoreConstants.RealmConfig.PROPERTY_DOMAIN_NAME);
+            }
+        }
+        return userStoreDomainName;
```

Comment:
I don't think it's safe to populate the user store domain like this. Ideally as the events triggers that specific point should know what the user store domain is.
If user store domain is not available don't add it to the payload

Reply:
Fixed - 17a7c015a1908ba70749fb300ec30f27cae4af17